omnibus/files/private-chef-cookbooks/private-chef/libraries/preflight_postgres_validator.rb

```diff
@@ -0,0 +1,322 @@
+      case e.message
+      when /.*Connection refused.*/
+        fail_with err_postgres_not_available
+      when /.*password authentication failed.*/
+        # This is what we want to see.
+      when /.*no pg_hba.conf entry.*/
+        # This is also possible, depending on if they've set up pg_hba
+        # by host or user or both. For common validation of invalid user - right
+        # now we just want to ensure that we connected.
+      else
+        # None of these should be happening, but we still don't want to dump an
+        # unhelpful stack trace on the screen so let's catch it and pass the message along.
+        fail_with "#{CSPG999}: #{e.message}"
+      end
+    end
+  end
+
+  def backend_validation
+    begin
+      connect_as(:superuser) do |connection|
+        backend_verify_database_access(connection)
+        backend_verify_postgres_version(connection)
+        %w{bifrost opscode_chef oc_id}.each {|db| backend_verify_empty_db(connection, db) }
+      end
+    rescue ::PG::InsufficientPrivilege => e
+      fail_with err_not_superuser
+    rescue ::PG::ConnectionBad => e
+      case e.message
+      when /.*database 'template1', does not exist.*/
+        fail_with err_template1_db_required
+      when /.*password authentication failed.*/
+        fail_with err_invalid_superuser_account
+      when /.*no pg_hba.conf entry.*/
+        fail_with err_invalid_pg_hba
+      else
+        # None of these should be happening, but we still don't want to dump an
+        # unhelpful stack trace on the screen so let's catch it and pass the message along.
+        fail_with "#{CSPG999}: #{e.message}"
+      end
+    end
+  end
+
+
+
+private
+  def backend_verify_database_access(connection)
+    # Make sure we have the access we need.
+    # TODO - this may require changes for RDS support and/or restricted environments.
+    # In particular we may not have access to this information in this way -
+    # look at pg_has_role, perhaps? or has_database_privilege?
+    # Note on not escaping username: we already connected with the same attribute,
+    # so it's safe at this point.
+    r = connection.exec("SELECT rolsuper, rolcreaterole, rolcreatedb FROM pg_authid WHERE rolname='#{cs_pg_attr['db_superuser']}';")
+    # a super user may not have createrole/createdb flags set, so check for both cases
+    unless (r[0]['rolsuper'] == 't') or (r[0]['rolcreaterole'] == 't' and r[0]['rolcreatedb'] == 't')
+      fail_with err_not_superuser
+    end
+  end
+
+  def backend_verify_postgres_version(connection)
+    # Make sure the server is a supported version.
+    r = connection.exec("SHOW server_version;")
+    v = r[0]['server_version']
+    major, minor = v.split(".")
+    # Load up our required major/minor:
+    # NOTE: our current entry in version-manifest.json is 'postgres92',effectively hardcoding the version.
+    # TODO: We will want to ensure that current postgres is always labeled as 'postgres' which will have migration
```

Comment:
We have no way of turning a TODO here into an action.  Let's put this somewhere else.  Perhaps as an issue against chef-server.